archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,252 @@
+/-
+Copyright (c) 2021 Mantas Bakšys. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mantas Bakšys
+-/
+
+import data.real.sqrt
+import tactic.interval_cases
+import tactic.linarith
+import tactic.norm_cast
+import tactic.norm_num
+import tactic.ring_exp
+
+
+/-!
+# IMO 2021 Q1
+Let n≥100 be an integer. Ivan writes the numbers n, n+1,..., 2n each on different cards.
+He then shuffles these n+1 cards, and divides them into two piles. Prove that at least one
+of the piles contains two cards such that the sum of their numbers is a perfect square.
+# Solution
+We show there exists a triplet a, b, c ∈ [n , 2n] with a < b < c and each of the sums (a + b),
+(b + c), (a + c) being a perfect square. Specifically, we consider the linear system of equations
+
+  a + b = (2 * l - 1) ^ 2
+  a + c = (2 * l) ^ 2
+  b + c = (2 * l + 1) ^ 2
+
+which can be solved to give
+
+  a = 2 * l * l - 4 * l
+  b = 2 * l * l + 1
+  c = 2 * l * l + 4 * l
+
+Therefore, it is enough to show that there exists a natural number l such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 100.
+
+Then, by Pigeonhole principle, at least two numbers in the triplet must lie in the same pile, which
+finishes the proof.
+-/
+
+
+open real
+
+lemma lower_bound (n l : ℕ) (hl : 2 + sqrt (4 + 2 * n) ≤ 2 * (l : ℝ)) :
+  (n : ℕ) + 4 * l ≤ 2 * l * l :=
+begin
+  have h₁ : sqrt (4 + 2 * n) ≤ 2 * l - 2, { linarith },
+  replace h₁ := (sqrt_le_iff.1 h₁).2,
+  ring_exp at h₁,
+  norm_num at h₁,
+  norm_cast at h₁,
+  linarith,
+end
+
+lemma upper_bound (n l : ℕ) (hl₂ : (l : ℝ) ≤ sqrt (1 + n) - 1) :
+  2 * l * l + 4 * l ≤ 2 * n :=
+begin
+  have h₁ : (l : ℝ) + 1 ≤ sqrt (1 + n), { linarith },
```

Comment:
It's not essential, but when `linarith` is just doing a single obvious step some people prefer to use the library lemma, which you can typically find using, e.g.:
```
example (a b : ℤ) (h : a ≤ b - 1) : a + 1 ≤ b := by library_search
```